use fs2 for TextInput, TextOutput, BinaryInput

jenshalm opened the pull request:
This is the main migration step.
    
Adjusting `TextInput` and `TextOutput` is straightforward due to the simplification in the first step.
    
`BinaryInput` will receive a breaking change by modelling the input as `fs2.Stream[F, Byte]`. This is also "public-internal" API with little impact on end users.

`BinaryOutput` is the unfortunate omission here: it cannot be migrated to fs2. The reason is that the model is currently solely used for binary output formats like PDF and EPUB which require the use of Java APIs that would not know how to handle an `fs2.Pipe`. This leaves an odd inconsistency between binary input and output, but half a migration is still better than none. This also has an impact on potential, future support for Scala.js in `laika-io` which is discussed further below.

The types `InputRuntime`, `OutputRuntime` and `CopyRuntime` in the `laika.io.runtime` package can be removed entirely. Now that fs2 is the IO runtime they are not needed and the little integration and plumbing that is required stays in the companions of the input and output model. Additional benefit is that the amount of code to maintain in `laika-io` shrinks by 200 lines of code.

Another integration challenge is the `FopResourceResolver` for Apache FOP in the PDF renderer, which provides font, image and other resources to FOP. Here we need to convert an `fs2.Stream` back to a Java `InputStream` to implement a synchronous Java callback. This is mildly inefficient, but given that PDF has by far the slowest renderer in Laika anyway, the difference should not be noticeable.

Final oddity is the `ContainerWriter` for EPUB output which now talks to a `ZipOutputStream` and an `fs2.Pipe` pointing to the same underlying `OutputStream` to fulfil differing integration requirements. At least in the test suites that works fine.

Some further manual testing is required before this can be merged, as some functionality cannot be easily checked by Laika's test suites:

- [ ] visual checks for EPUB output
- [ ] visual checks for PDF output
- [ ] working with the preview server

armanbilge commented on io/src/main/scala/laika/io/model/Input.scala:
```diff
@@ -71,17 +72,21 @@ case class TextInput[F[_]: Functor] (path: Path, docType: TextDocumentType, inpu
 
-  private def readAll[F[_]: Sync](reader: Resource[F, Reader], sizeHint: Int): F[String] = 
-   reader.use(InputRuntime.readAll(_, sizeHint))
+  private def readAll[F[_]: Sync](input: fs2.Stream[F, Byte], codec: Codec): F[String] = 
+    input.through(fs2.text.decodeWithCharset(codec.charSet)).compile.string
```
```suggestion
  private def readAll[F[_]: Concurrent](input: fs2.Stream[F, Byte], codec: Codec): F[String] = 
    input.through(fs2.text.decodeWithCharset(codec.charSet)).compile.string
```

FS2 has both a `Concurrent` compiler and a `Sync` compiler, but using the `Sync` compiler for a potentially `Concurrent` stream is not ideal (http4s has an internal lint to prevent this "downgrade"). See discussion in https://github.com/typelevel/fs2/pull/2657.

jenshalm replied:
Oh, that is quite subtle and somewhat counterintuitive - something I wasn't aware of. I also could not find references to this in the fs2 docs.

armanbilge replied:
Yeah you are absolutely right 😕 I guess it's sort of documented here:
https://github.com/typelevel/fs2/blob/9db8a220a07fc5b203eff4d311f53f769fb66a5d/core/shared/src/main/scala/fs2/Compiler.scala#L54

armanbilge replied:
(But please open an issue about how we might improve that.)

armanbilge commented on io/src/main/scala/laika/io/runtime/VersionedLinkTargets.scala:
```diff
@@ -72,9 +72,10 @@ private[runtime] object VersionedLinkTargets {
-      .use(InputRuntime.readAll(_, 8096))
+    existing.input
+      .through(fs2.text.utf8.decode)
+      .compile
```
`Sync` compiler being used here I think.

armanbilge commented on io/src/main/scala/laika/io/runtime/RendererRuntime.scala:
```diff
@@ -128,6 +128,10 @@ object RendererRuntime {
       val pathTranslator = createPathTranslator(translatorConfig, Root / "dummy", lookup).translate(_:Path)
+
+      def createDirectory (file: File): F[Unit] =
+        Sync[F].delay(file.exists || file.mkdirs()).flatMap(if (_) Sync[F].unit
```
```suggestion
        Sync[F].blocking(file.exists || file.mkdirs()).flatMap(if (_) Sync[F].unit
```

Why not use `Files[F]` here?

jenshalm replied:
This part gets migrated in the next PR where creating and scanning directories is handled. Leaving this as it is, for a smoother merge.

armanbilge commented on io/src/main/scala/laika/render/epub/ZipWriter.scala:
```diff
@@ -37,19 +35,18 @@ object ZipWriter {
     * a generic zip utility as the method name suggests.
     */
-  def zipEPUB[F[_]: Sync] (inputFs: Seq[BinaryInput[F]], output: BinaryOutput[F]): F[Unit] = {
+  def zipEPUB[F[_]: Sync] (inputs: Seq[BinaryInput[F]], output: BinaryOutput[F]): F[Unit] = {
```
I wonder if some of this `Zip` stuff could be upstreamed into FS2. I think it came up on the Discord the other day.

jenshalm replied:
That would be an interesting API design challenge, given how much metadata you'd need to route through such a stream. Also note this funny requirement that EPUB has: the first entry needs to be uncompressed. The API would need to support these kind of edge cases.

armanbilge commented on io/src/main/scala/laika/render/epub/ZipWriter.scala:
```diff
@@ -37,19 +35,18 @@ object ZipWriter {
         _     <- prepareEntry(entry)
         _     <- Sync[F].blocking(zipOut.putNextEntry(entry))
-        _     <- CopyRuntime.copy(input._1, zipOut)
+        _     <- input.input.through(pipe).compile.drain
```
Sync compiler again I suspect.

armanbilge commented on io/src/main/scala/laika/render/epub/ZipWriter.scala:
```diff
@@ -60,18 +57,14 @@ object ZipWriter {
 
       writeEntry(inputs.head, prepareUncompressedEntry) >> 
-        inputs.tail.map(writeEntry(_)).sequence >> 
+        inputs.toList.tail.map(writeEntry(_)).sequence >> 
```
```suggestion
        inputs.toList.tail.traverse(writeEntry(_)) >> 
```

armanbilge commented on io/src/main/scala/laika/helium/generate/MergedCSSGenerator.scala:
```diff
@@ -16,13 +16,19 @@
 
+  private def merge[F[_]: Sync](inputs: Seq[BinaryInput[F]]): F[String] =
+    inputs.toList
+      .map(_.input.through(fs2.text.utf8.decode).compile.string)
+      .sequence
```
```suggestion
      .traverse(_.input.through(fs2.text.utf8.decode).compile.string)
```

Also note the sync compiler used here.

jenshalm commented on io/src/main/scala/laika/io/model/Input.scala:
```diff
@@ -16,44 +16,45 @@
   def fromStream[F[_]: Sync] (path: Path, stream: F[InputStream], autoClose: Boolean, targetFormats: TargetFormats = TargetFormats.All): BinaryInput[F] = {
-    val resource = if (autoClose) Resource.fromAutoCloseable(stream) else Resource.eval(stream)
-    BinaryInput(path, resource, targetFormats)
+    val input = fs2.io.readInputStream(stream, 8096, autoClose)
```
True, I found traces of either 32K or 64K as defaults in the code base. `Files[F].readAll` uses 64K, so I could use that here simply for consistency. I would assume (hope) the fs2 devs arrived at these values via some benchmarking. 🙂

jenshalm commented:
Thanks a lot, that was very helpful. In particular, I wasn't aware of those fs2 compiler subtleties. And apparently `traverse` is not as ingrained in my memory as in the "the answer is always traverse" camp. 😆 

I addressed all comments except for one which is deferred to the next PR.

armanbilge commented on io/src/main/scala/laika/helium/generate/MergedCSSGenerator.scala:
```diff
@@ -16,13 +16,18 @@
 
 private[helium] object MergedCSSGenerator {
 
+  private def merge[F[_]: Sync](inputs: Seq[BinaryInput[F]]): F[String] =
```
```suggestion
  private def merge[F[_]: Concurrent](inputs: Seq[BinaryInput[F]]): F[String] =
```

armanbilge commented on io/src/test/scala/laika/io/helper/ResultExtractor.scala:
```diff
@@ -33,14 +28,8 @@ trait ResultExtractor {
-      }
-    }
+    private def readText (stream: fs2.Stream[F, Byte]): F[String] = 
+      stream.through(fs2.text.utf8.decode).compile.string
```
Another sync compiler, although this is test sources.

armanbilge reviewed (commented):
Glad to have been of help :) yeah the compiler stuff is unfortunately subtle :( I checked through again and I think there's one last one.
